sbt-scalajs-bundler/src/main/scala/scalajsbundler/PackageManager.scala

```diff
@@ -0,0 +1,163 @@
+
+import sbt._
+import scalajsbundler.util.Commands
+import scalajsbundler.util.JSON
+
+abstract class PackageManager {
+
+  val name: String
+  val installCommand: String
+  val installArgs: Seq[String]
+
+  /**
+    * Runs the command `cmd`
+    * @param args Command arguments
+    * @param workingDir Working directory of the process
+    * @param logger Logger
+    */
+  protected def run(args: String*)(workingDir: File, logger: Logger): Unit =
+    Commands.run(cmd ++: args, workingDir, logger)
+
+  private val cmd = sys.props("os.name").toLowerCase match {
+    case os if os.contains("win") => Seq("cmd", "/c", name)
+    case _                        => Seq(name)
+  }
+
+  def install(baseDir: File,
+              installDir: File,
+              logger: Logger): Unit = {
+    this match {
+      case lfs: LockFileSupport =>
+        lfs.lockFileRead(baseDir, installDir, logger)
+      case _ =>
+        ()
+    }
+
+    run(installCommand +: installArgs: _*)(installDir, logger)
+
+    this match {
+      case lfs: LockFileSupport =>
+        lfs.lockFileWrite(baseDir, installDir, logger)
+      case _ =>
+        ()
+    }
+  }
+
+  val packageJsonContents: Map[String, JSON]
+}
+
+trait AddPackagesSupport { this: PackageManager =>
+
+  val addPackagesCommand: String
+  val addPackagesArgs: Seq[String]
+
+  /**
+    * Locally install NPM packages
+    *
+    * @param baseDir The (sub-)project directory which contains yarn.lock
+    * @param installDir The directory in which to install the packages
+    * @param logger sbt logger
+    * @param npmPackages Packages to install (e.g. "webpack", "webpack@2.2.1")
+    */
+  def addPackages(baseDir: File,
+                  installDir: File,
+                  logger: Logger,
+                 )(npmPackages: String*): Unit = {
+    this match {
+      case lfs: LockFileSupport  =>
+        lfs.lockFileRead(baseDir, installDir, logger)
+      case _ =>
+        ()
+    }
+
+    run(addPackagesCommand +: (addPackagesArgs ++ npmPackages): _*)(installDir, logger)
+
+    this match {
+      case lfs: LockFileSupport  =>
+        lfs.lockFileWrite(baseDir, installDir, logger)
+      case _ =>
+        ()
+    }
+  }
+}
+
+trait LockFileSupport {
+  val lockFileName: String
+
+  def lockFileRead(
+    baseDir: File,
+    installDir: File,
+    logger: Logger
+  ): Unit = {
+    val sourceLockFile = baseDir / lockFileName
+    val targetLockFile = installDir / lockFileName
+
+    if (sourceLockFile.exists()) {
+      logger.info("Using lockfile " + sourceLockFile)
+      IO.copyFile(sourceLockFile, targetLockFile)
+    }
+  }
+
+  def lockFileWrite(
+    baseDir: File,
+    installDir: File,
+    logger: Logger
+  ): Unit = {
+    val sourceLockFile = baseDir / lockFileName
+    val targetLockFile = installDir / lockFileName
+
+    if (targetLockFile.exists()) {
+      logger.debug("Wrote lockfile to " + sourceLockFile)
+      IO.copyFile(targetLockFile, sourceLockFile)
+    }
+  }
+}
+
+case class Npm(
```

Comment:
Never use `case class`es for things that you don't intend to `match/case` on. These are impossible to evolve in binary-compatible ways.

Use a `final class` with a `Config` parameter instead, where `Config` follows the pattern that you can see at
https://github.com/scala-js/scala-js-js-envs/blob/9d46d49310619de3d722b5d3867fa7599b1a5831/nodejs-env/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala#L213-L270
Yes, it is a pain to write, but it is the only way not to completely lock this down for the future.